Source/GSXib5KeyedUnarchiver.m

```diff
@@ -0,0 +1,3391 @@
+                                                            andAttributes: [NSDictionary dictionaryWithObjectsAndKeys:
+                                                                                           @"IBUserDefinedRuntimeAttributesPlaceholder", @"class",
+                                                                                         @"IBUserDefinedRuntimeAttributesPlaceholderName", @"key",
+                                                                                         nil]];
+      GSXibElement *placeStringRec = [[GSXibElement alloc] initWithType: @"string"
+                                                            andAttributes: [NSDictionary dictionaryWithObjectsAndKeys:
+                                                                                           @"name", @"key",
+                                                                            nil]];
+      GSXibElement *referenceRec   = [self createReference: [element attributeForKey: @"id"]];
+
+      // Setup placeholder record...
+      [placeStringRec setValue: @"IBUserDefinedRuntimeAttributesPlaceholderName"];
+      [placeholderRec setElement: placeStringRec forKey: @"name"];
+      [placeholderRec setElement: referenceRec forKey: @"object"];
+      [placeholderRec setElement: element forKey: @"userDefinedRuntimeAttributes"];
+
+      // Attach placeholder and string records to the object record...
+      [objectRecord setElement: stringRecord forKey: @"NS.key.0"];
+      [objectRecord setElement: placeholderRec forKey: @"IBUserDefinedRuntimeAttributesPlaceholderName"];
+
+      // Add to flattened properties...
+      [_flattenedProperties setElement: objectRecord forKey: refID];
+
+      // Cleanup...
+      RELEASE(stringRecord);
+      RELEASE(placeholderRec);
+      RELEASE(placeStringRec);
+      RELEASE(objectRecord);
+    }
+}
+
+- (id) initForReadingWithData: (NSData*)data
+{
+  // Check data...
+  if (data == nil)
+    {
+      DESTROY(self);
+      return nil;
+    }
+#if     GNUSTEP_BASE_HAVE_LIBXML
+  else
+    {
+      // Ensure we have a XIB 5 version...first see if we can parse the XML...
+      NSXMLDocument *document = [[NSXMLDocument alloc] initWithData: data
+                                                            options: 0
+                                                              error: NULL];
+      if (document == nil)
+        {
+          NSLog(@"%s:DOCUMENT IS NIL: %@\n", __PRETTY_FUNCTION__, document);
+          DESTROY(self);
+          return nil;
+        }
+      else
+        {
+          // Test to see if this is an Xcode 5 XIB...
+          NSArray *documentNodes = [document nodesForXPath: @"/document" error: NULL];
+
+          // Need at LEAST ONE document node...we should find something a bit more
+          // specific to check here...
+          if ([documentNodes count] == 0)
+            {
+              NSWarnMLog(@"XIB is NOT a version 5 level");
```

Comment:
So, what is the option that the user has here -- try again with a different deserializer? Isn't this a problem if you can't tell the version from the extension? Would it make sense to transparently downgrade to pre-XIB5 deserializer?

Reply:
That is what happens within GSXibLoader.